##### Joshua Boehm
##### 26 September 2022
##### MATH 3423

# Libraries

In [7]:
import numpy as np
import scipy as sp

# Preface

Let A be a (5 x 3) randomly generated matrix `A = np.random.randint( -2, 5, size = (5, 3) )` with integer elements in the interval [-2, 5); and let b a (5 x 1) randomly generated vector `b = np.random.randint( -2, 3, size = (5, 1))` with integer elements in the interval [-2, 3).

In [8]:
# Assigning the randomly generated matrices to variables in order to manipulate below without mutating them.
a = np.random.randint(-2,5, size = (5,3))
B = np.random.randint(-2,3, size = (5,1))

# I randomized the above code until I got a matrix that worked with Cholesky's (both pre and post-multiplied)
# These are those matrices. If you like, comment them to re-randomize the matrices.
a = np.array([
    [4,0,2],
    [1,2,0],
    [4,1,-2],
    [0,0,3],
    [2,4,4]],float)
B = np.array([
    [0],
    [-1],
    [2],
    [0],
    [0]],float)

In [9]:
Aposdef = np.array([
    [4,0,2],
    [1,2,0],
    [4,1,-2],
    [0,0,3],
    [2,4,4]],float)
bposdef= np.array([
    [0],
    [-1],
    [2],
    [0],
    [0]],float)
print(a)
print(bposdef)
a-Aposdef
B-bposdef

[[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]]
[[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]]


array([[0.],
       [0.],
       [0.],
       [0.],
       [0.]])

# Question 1
Find the LU factorization of $A^TA$ and $A A^T$ and use each factorization to solve: 
$$ A^TAx = A^Tb \quad{and}\quad AA^Tx = b$$


## Solution

### A<sup>T</sup>A variant

First, we examine pre-multiplying $A$ with $A^T$, resulting in a square matrix of size $3 \times 3$. As a result, we will evaluate the problem $A^TAx = A^Tb$

In [10]:
from scipy.linalg import lu
from scipy.linalg import solve

A = a.T @ a
b = a.T @ B

P, L, U = lu(A)

print(f" \
The original A matrix: \n {a} \n\n \
The original right-side matrix b: \n {B} \n\n \
The matrix A^t A: \n {A} \n\n \
The right-side matrix A^t b \n {b} \n\n \
The Permutation matrix: \n {P} \n\n \
The Lower matrix: \n {L} \n\n \
The Upper matrix: \n {U} \n")




print("Reconstruct A with Factorization")
print(P @ L @ U)

# Solve A x = b
print(" ---> 1st: Solve L y = P^t b --> For the (m x 1) Vector y ")
y = solve(L, P.T @ b)
print(y)
print("\n ---> 2nd: Solve U x = y --> x ")
x = solve(U, y)
print(x)
x = solve(A, b)
print("\n ---> The Solution Vector x from solve:")
print(x)
print(" ")
print(" ---> Reconstruct the Vector P L U x --> A^t b")
print(P @ L @ U @ x)
print(" ")
print("A^t b matrix for comparison:")
print(b)

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The original right-side matrix b: 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 The matrix A^t A: 
 [[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]] 

 The right-side matrix A^t b 
 [[ 7.]
 [ 0.]
 [-4.]] 

 The Permutation matrix: 
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]] 

 The Lower matrix: 
 [[1.         0.         0.        ]
 [0.37837838 1.         0.        ]
 [0.21621622 0.69879518 1.        ]] 

 The Upper matrix: 
 [[37.         14.          8.        ]
 [ 0.         15.7027027  10.97297297]
 [ 0.          0.         23.60240964]] 

Reconstruct A with Factorization
[[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]]
 ---> 1st: Solve L y = P^t b --> For the (m x 1) Vector y 
[[ 7.        ]
 [-2.64864865]
 [-3.6626506 ]]

 ---> 2nd: Solve U x = y --> x 
[[ 0.24553344]
 [-0.06023481]
 [-0.15518121]]

 ---> The Solution Vector x from solve:
[[ 0.24553344]
 [-0.06023481]
 [-0.15518121]]
 
 ---> 

The resulting solution vector $x$ we get is $A^{T}b$ is a $3 \times 1$ matrix. This is a consequence of $A^T$ being a $3 \times 5$ matrix multiplied by $b$, a $5 \times 1$ matrix. I would explicitly write it in Latex, except that it is randomly generated and will remain so unless I happen upon a Positive-Definite matrix (to use in Cholesky's).

__**EDIT:**__ After finding a positive-definite matrix, I have saved it to be used with all of the homework questions (including this one). The solve does prove to be correct by reconstruction if garbage float-point 0 is considered.

That matrix $x$ is as follows: 
$$x = \begin {bmatrix}
0.24553344 \\
-0.06023481 \\
-0.15518121
\end {bmatrix}$$

### AA<sup>T</sup> variant

Here, we post-multiply $A$ with $A^T$ giving us a square matrix with the diminisions $5 \times 5$, which we will utilize to evalute the equation $AA^Tx = b$.

In [11]:
A = a @ a.T
P, L, U = lu(A)
b = B

print(f" \
The matrix AA^t: \n {A} \n\n \
The Permutation matrix: \n {P} \n\n \
The Lower matrix: \n {L} \n\n \
The Upper matrix: \n {U.round(2)} \n\n \
The right-side matrix b \n {b} \n")


print(" ---> 1st: Solve L y = P^t b --> For the (m x 1) Vector y ")
y = solve(L, P.T @ b)
print(y)
print(" ")
print(" ---> 2nd: Solve U x = y --> x ")
x = solve(U, y)
print(x)
x = solve(A, b)
print(" ")
print(" ---> The (n x 1) Solution Vector x from solve:")
print(x)
print(" ")
print(" ---> Reconstruct the (m x 1) Vector P L U x --> b ")
print(P @ L @ U @ x)
print(" ")
print("Original b matrix for comparison:")
print(b)

 The matrix AA^t: 
 [[20.  4. 12.  6. 16.]
 [ 4.  5.  6.  0. 10.]
 [12.  6. 21. -6.  4.]
 [ 6.  0. -6.  9. 12.]
 [16. 10.  4. 12. 36.]] 

 The Permutation matrix: 
 [[1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 1. 0. 0. 0.]] 

 The Lower matrix: 
 [[ 1.          0.          0.          0.          0.        ]
 [ 0.8         1.          0.          0.          0.        ]
 [ 0.6         0.52941176  1.          0.          0.        ]
 [ 0.3        -0.17647059 -0.63157895  1.          0.        ]
 [ 0.2         0.61764706  0.42105263 -0.17737948  1.        ]] 

 The Upper matrix: 
 [[ 20.     4.    12.     6.    16.  ]
 [  0.     6.8   -5.6    7.2   23.2 ]
 [  0.     0.    16.76 -13.41 -17.88]
 [  0.     0.     0.    -0.    -0.  ]
 [  0.     0.     0.     0.    -0.  ]] 

 The right-side matrix b 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 ---> 1st: Solve L y = P^t b --> For the (m x 1) Vector y 
[[ 0.        ]
 [ 0.        ]
 [ 2.        ]
 [ 1.26315789]
 [-1.

/var/folders/90/b16ybj8s2gv__lhxf5ttmvj40000gn/T/ipykernel_29510/102300893.py:18: LinAlgWarning: Ill-conditioned matrix (rcond=1.24777e-18): result may not be accurate.
  x = solve(U, y)
/var/folders/90/b16ybj8s2gv__lhxf5ttmvj40000gn/T/ipykernel_29510/102300893.py:20: LinAlgWarning: Ill-conditioned matrix (rcond=9.13148e-19): result may not be accurate.
  x = solve(A, b)


This matrix turned out to be ill-conditioned and as a result, the solution $x$ turned out to be unusable. When plugging the solution $x$ into the matrix $AA^T$, the right side $b$ is not the same as the original.

# Question 2
Find the $LDL^T$ factorization of $A^TA$ and $A A^T$ and use each factorization to solve: 
$$ A^TAx = A^Tb \quad{and}\quad AA^Tx = b$$

## Solution

### A<sup>T</sup>A variant

Again, we examine pre-multiplying $A$ with $A^T$, resulting in a square matrix of size $3 \times 3$. As a result, we will evaluate the problem $A^TAx = A^Tb$.

In [20]:
from scipy.linalg import ldl
from scipy.linalg import solve, det

A = a.T @ a
b = a.T @ B

L, D, P = ldl(A)


print(f" \
The original A matrix: \n {a} \n\n \
The original right-side matrix b: \n {B} \n\n \
The matrix A^t A: \n {A} \n\n \
The right-side matrix A^t b \n {b} \n\n \
The Permutation matrix: \n {P} \n\n \
The Lower matrix: \n {L} \n\n \
The Diagonal matrix: \n {D} \n")

print("----- Reconstruct A = L D L^t ----- ")
print(L @ D @ L.T)

print(" -----  1st: Solve L z = b  for z ----- ")
z = solve(L, b)
print(z)
print(" ----- 2nd: Solve D y = z  for y ----- ")
y = solve(D, z)
print(y)
print(" ----- 3rd: Solve L^t x = y  for x ----- ")
x = solve(L.T, y)
print(x)
print(" ----- Reconstruct the (m x 1) b Vector")
print(" ----- Is  L @ D @ L.T @ x = b =\n",  L @ D @ L.T @ x )
print("The original A^t b matrix:\n",b)

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The original right-side matrix b: 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 The matrix A^t A: 
 [[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]] 

 The right-side matrix A^t b 
 [[ 7.]
 [ 0.]
 [-4.]] 

 The Permutation matrix: 
 [0 1 2] 

 The Lower matrix: 
 [[1.         0.         0.        ]
 [0.37837838 1.         0.        ]
 [0.21621622 0.69879518 1.        ]] 

 The Diagonal matrix: 
 [[37.          0.          0.        ]
 [ 0.         15.7027027   0.        ]
 [ 0.          0.         23.60240964]] 

----- Reconstruct A = L D L^t ----- 
[[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]]
 -----  1st: Solve L z = b  for z ----- 
[[ 7.        ]
 [-2.64864865]
 [-3.6626506 ]]
 ----- 2nd: Solve D y = z  for y ----- 
[[ 0.18918919]
 [-0.1686747 ]
 [-0.15518121]]
 ----- 3rd: Solve L^t x = y  for x ----- 
[[ 0.24553344]
 [-0.06023481]
 [-0.15518121]]
 ----- Reconstruct the (m x 1) b Vector


As we can see, the $LDL^T$ decomposition and solution works as equally well as the $LU$ decomposition did above for this $A^TAx = A^TB$ equation.

### AA<sup>T</sup> variant

Again, we post-multiply $A$ with $A^T$ giving us a square matrix with the diminisions $5 \times 5$, which we will utilize to evalute the equation $AA^Tx = b$. Hopefully this decomposition is less prone to the rounding errors we ran into above.

In [21]:
from scipy.linalg import ldl
from scipy.linalg import solve, det

A = a @ a.T
b = B

L, D, P = ldl(A)

print(f" \
The original A matrix: \n {a} \n\n \
The matrix A A^t: \n {A} \n\n \
The original right-side matrix b: \n {b} \n\n \
The Lower matrix: \n {L} \n\n \
The Diagonal matrix: \n {D.round(2)} \n")
#The Permutation matrix: \n {P} \n\n \

print("----- Reconstruct A = L D L^t ----- ")
print(L @ D @ L.T)

print(" -----  1st: Solve L z = b  for z ----- ")
z = solve(L, b)
print(z)
print(" ----- 2nd: Solve D y = z  for y ----- ")
y = solve(D, z)
print(y)
print(" ----- 3rd: Solve L^t x = y  for x ----- ")
x = solve(L.T, y)
print(x)
print(" ----- Reconstruct the (m x 1) b Vector")
print(" ----- Is  L @ D @ L.T @ x = b =\n",  L @ D @ L.T @ x )
print("The original b matrix:\n",b)

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The matrix A A^t: 
 [[20.  4. 12.  6. 16.]
 [ 4.  5.  6.  0. 10.]
 [12.  6. 21. -6.  4.]
 [ 6.  0. -6.  9. 12.]
 [16. 10.  4. 12. 36.]] 

 The original right-side matrix b: 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 The Lower matrix: 
 [[ 1.          0.          0.          0.          0.        ]
 [ 0.2         1.          0.          0.          0.        ]
 [ 0.6         0.85714286  1.          0.          0.        ]
 [ 0.3        -0.28571429 -0.8         1.          0.        ]
 [ 0.8         1.61904762 -1.06666667  0.84848485  1.        ]] 

 The Diagonal matrix: 
 [[20.    0.    0.    0.    0.  ]
 [ 0.    4.2   0.    0.    0.  ]
 [ 0.    0.   10.71  0.    0.  ]
 [ 0.    0.    0.   -0.    0.  ]
 [ 0.    0.    0.    0.    0.  ]] 

----- Reconstruct A = L D L^t ----- 
[[ 2.00000000e+01  4.00000000e+00  1.20000000e+01  6.00000000e+00
   1.60000000e+01]
 [ 4.00000000e+00  5.00000000

Yet again, we have the problem of an ill-conditioned matrix creating wild rounding errors. I'm a little concerned going into the future with this.

# Question 3
Find the $Cholesky$ factorization of $A^TA$ and $A A^T$ and use each factorization to solve: 
$$ A^TAx = A^Tb \quad{and}\quad AA^Tx = b$$

## Solution

In order for us to even utilize $Cholesky's$, our factorable matrix must be positive-definite.

### A<sup>T</sup>A variant

In [44]:
from numpy import array
from numpy.linalg import cholesky
from scipy.linalg import solve

A = a.T @ a
b = a.T @ B

L = cholesky(A)
y = solve(L, b)
x = solve(L.T, y)

print(f" \
The original A matrix: \n {a} \n\n \
The matrix A^t A: \n {A} \n\n \
The original right-side matrix b: \n {B} \n\n \
The right-side matrix A^t b: \n {b} \n\n \
The Cholesky Lower matrix: \n {L} \n\n \
Reconstruction of LL^t (should be A^t A): \n {L @ L.T} \n \
Solve A x = L L^t x =  b \
1st: Solve L y =  b --> y \n {y} \n\n \
2nd: Solve L^t x = y --> x \n {x} \n\n \
Is L L^t x = A^t b ? \n {L @ L.T @ x}\n\n \
The A^t b matrix for comparison: \n {b}")

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The matrix A^t A: 
 [[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]] 

 The original right-side matrix b: 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 The right-side matrix A^t b: 
 [[ 7.]
 [ 0.]
 [-4.]] 

 The Cholesky Lower matrix: 
 [[6.08276253 0.         0.        ]
 [2.30158582 3.96266359 0.        ]
 [1.3151919  2.76909022 4.85823112]] 

 Reconstruction of LL^t (should be A^t A): 
 [[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]] 
 Solve A x = L L^t x =  b 1st: Solve L y =  b --> y 
 [[ 1.15079291]
 [-0.66840109]
 [-0.75390621]] 

 2nd: Solve L^t x = y --> x 
 [[ 0.24553344]
 [-0.06023481]
 [-0.15518121]] 

 Is L L^t x = A^t b ? 
 [[ 7.00000000e+00]
 [ 1.66533454e-16]
 [-4.00000000e+00]]

 The A^t b matrix for comparison: 
 [[ 7.]
 [ 0.]
 [-4.]]


As above with the $A^TA$ matrix, the solve turns out as expected. The only real difference is the display of 0 as a floating point number.

### AA<sup>T</sup> variant

Once more, we post-multiply $A$ with $A^T$ giving us a square matrix with the diminisions $5 \times 5$, which we will utilize to evalute the equation $AA^Tx = b$. Perhaps this will perform better than the $LDL^T$ decomposition.

In [45]:
from numpy import array
from numpy.linalg import cholesky
from scipy.linalg import solve

A = a @ a.T
b = B

L = cholesky(A)
y = solve(L, b)
x = solve(L.T, y)

print(f" \
The original A matrix: \n {a} \n\n \
The matrix A A^t: \n {A} \n\n \
The original right-side matrix b: \n {B} \n\n \
The Cholesky Lower matrix: \n {L.round(2)} \n\n \
Reconstruction of LL^t (should be A A^t): \n {(L @ L.T).round(2)} \n \
Solve A x = L L^t x =  b \
1st: Solve L y =  b --> y \n {y} \n\n \
2nd: Solve L^t x = y --> x \n {x} \n\n \
Is L L^t x = b ? \n {L @ L.T @ x}\n\n \
The A^t b matrix for comparison: \n {b}")

np.linalg.cond(A)

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The matrix A A^t: 
 [[20.  4. 12.  6. 16.]
 [ 4.  5.  6.  0. 10.]
 [12.  6. 21. -6.  4.]
 [ 6.  0. -6.  9. 12.]
 [16. 10.  4. 12. 36.]] 

 The original right-side matrix b: 
 [[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]] 

 The Cholesky Lower matrix: 
 [[ 4.47  0.    0.    0.    0.  ]
 [ 0.89  2.05  0.    0.    0.  ]
 [ 2.68  1.76  3.27  0.    0.  ]
 [ 1.34 -0.59 -2.62  0.    0.  ]
 [ 3.58  3.32 -3.49  0.    0.  ]] 

 Reconstruction of LL^t (should be A A^t): 
 [[20.  4. 12.  6. 16.]
 [ 4.  5.  6. -0. 10.]
 [12.  6. 21. -6.  4.]
 [ 6. -0. -6.  9. 12.]
 [16. 10.  4. 12. 36.]] 
 Solve A x = L L^t x =  b 1st: Solve L y =  b --> y 
 [[ 0.00000000e+00]
 [-4.87950036e-01]
 [ 8.72871561e-01]
 [ 4.74531328e+07]
 [ 4.08411715e+07]] 

 2nd: Solve L^t x = y --> x 
 [[-1.03277312e+15]
 [-1.55917684e+15]
 [ 1.18031213e+15]
 [ 8.29376323e+14]
 [ 4.84510384e+14]] 

 Is L L^t x = b ? 
 [[-4.        ]
 [-1.12

1.4461175500930304e+17

Yet again, every time I try to calculate the $AA^T$ variant of this problem, there are big issues. It seems one of two things must be happening:
+ I'm messing up, and the conclusions are invalid.
+ The bigger the matrix, the more likely it is to fail; use the smaller resulting square matrix.

# Question 4

Find the eigenvalues of the corresponding eigenvectors of $A^TA$ and $AA^T$. Use the $QR$ factorizations to find the eigenvalues and compare your answers.

# Solution

In [46]:
from numpy import array
from scipy.linalg import qr, solve

AtransposeA = a.T @ a
AAtranspose = a @ a.T

print(
f"The eiginvalues and corresponding eigenvectors for A^T A are:\n\
{np.linalg.eig(AtransposeA)[1].round(2)}\n\n\
The eiginvalues and corresponding eigenvectors for A A^T are:\n\
{np.linalg.eig(AAtranspose)[1].round(2)}")

The eiginvalues and corresponding eigenvectors for A^T A are:
[[-0.65 -0.69  0.32]
 [-0.5   0.08 -0.86]
 [-0.56  0.72  0.4 ]]

The eiginvalues and corresponding eigenvectors for A A^T are:
[[ 0.51  0.47 -0.68  0.25 -0.02]
 [ 0.22  0.37  0.46  0.1  -0.78]
 [ 0.27 -0.53  0.13  0.79  0.02]
 [ 0.23 -0.6  -0.39 -0.42 -0.49]
 [ 0.75 -0.05  0.39 -0.35  0.39]]


The above method utilizes the numpy libraries to find the eigenvalues and corresponding eigenvectors of $A^TA$ and $AA^T$.

In [47]:
# EigenValues using QR factorization

import numpy as np
from numpy.linalg import qr

# A1 = np.array([[0, 5, 1],
#                [1, 5, 2],
#                [1, 1, 4]], float)

MaxIter = 60; p = [1, 10, 20, MaxIter]
#A = A1

for i in range(MaxIter):
    Q, R = qr(A)
    A = R @ Q

    if i+1 in p:
        print(f"Iteration {i+1}:\n{R.round(2)}\n\n")

# print(f"A = \n{A}\n\n\
#     A1 = \n{A1}\n\n")

Iteration 1:
[[-29.19 -11.37 -18.64 -10.07 -36.18]
 [  0.    -6.9   -4.64   0.95 -12.53]
 [  0.     0.   -17.44  13.95  18.6 ]
 [  0.     0.     0.    -0.    -0.  ]
 [  0.     0.     0.     0.     0.  ]]


Iteration 10:
[[-5.476e+01 -4.000e-02  0.000e+00 -0.000e+00  0.000e+00]
 [ 0.000e+00 -2.695e+01  1.200e-01 -0.000e+00 -0.000e+00]
 [ 0.000e+00  0.000e+00 -9.290e+00 -0.000e+00  0.000e+00]
 [ 0.000e+00  0.000e+00  0.000e+00 -0.000e+00  0.000e+00]
 [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00  0.000e+00]]


Iteration 20:
[[-54.76  -0.     0.     0.     0.  ]
 [  0.   -26.95   0.    -0.     0.  ]
 [  0.     0.    -9.29   0.     0.  ]
 [  0.     0.     0.     0.     0.  ]
 [  0.     0.     0.     0.    -0.  ]]


Iteration 60:
[[-54.76   0.     0.     0.    -0.  ]
 [  0.   -26.95  -0.     0.    -0.  ]
 [  0.     0.     9.29   0.     0.  ]
 [  0.     0.     0.     0.    -0.  ]
 [  0.     0.     0.     0.    -0.  ]]




By 20 iterations, it seems to have converged sufficiently well. Perhaps, the interations are too high.

# Question 5

Find the $QR$ factorization of $A$, $A^TA$, and $AA^T$. Use these factorizations to solve $Ax =b$, $A^TAx = A^Tb$, and $AA^Tx = b$.

## Solution

In [64]:
import numpy as np
from scipy.linalg import qr
from numpy.linalg import solve

print(f"{a}\n\n{B}")
print(f"\n\
+--------------------------+ \n\
|  QR Factorization of A   | \n\
+--------------------------+ \n\
")
Q, R = qr(a, mode = "economic")
x = solve(R, Q.T @ B)
# Matrix A
print(f" \
The original A matrix: \n {a} \n\n \
The QR factorization:\n\n\
Q matrix:\n{Q.round(2)}\n\n\
R matrix:\n{R.round(2)}\n\n\
Solving for x by using QR x = b:\n\
{x}\n\n\n")


print(f"\n\
+--------------------------+ \n\
| QR Factorization of A^tA | \n\
+--------------------------+ \n\
")


Q, R = qr(a.T @ a)
# Matrix A^TA
print(f" \
The A^T A matrix: \n {a.T @ a} \n\n \
The QR factorization:\n\n\
Q matrix:\n{Q}\n\n\
R matrix:\n{R}\n\n\
Solving for x by using QR x = b:\n\
{solve(R, Q.T @ a.T @ B)}")

# A = AAtranspose
# b = B
Q, R = qr(a @ a.T)
x = solve (R, Q.T @ B)
print(f"\n\
+--------------------------+ \n\
| QR Factorization of AA^t | \n\
+--------------------------+ \n\
")

# Matrix AA^T
print(f" \
The A^T A matrix: \n {A} \n\n \
The QR factorization:\n\n\
Q matrix:\n{Q.round(2)}\n\n\
R matrix:\n{R.round(2)}\n\n\
Solving for x by using QR x = b:\n\
{solve(R, B).round(2)}")

[[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]]

[[ 0.]
 [-1.]
 [ 2.]
 [ 0.]
 [ 0.]]

+--------------------------+ 
|  QR Factorization of A   | 
+--------------------------+ 

 The original A matrix: 
 [[ 4.  0.  2.]
 [ 1.  2.  0.]
 [ 4.  1. -2.]
 [ 0.  0.  3.]
 [ 2.  4.  4.]] 

 The QR factorization:

Q matrix:
[[-0.66  0.38  0.45]
 [-0.16 -0.41 -0.28]
 [-0.66  0.13 -0.52]
 [-0.    0.    0.62]
 [-0.33 -0.82  0.27]]

R matrix:
[[-6.08 -2.3  -1.32]
 [ 0.   -3.96 -2.77]
 [ 0.    0.    4.86]]

Solving for x by using QR x = b:
[[ 0.24553344]
 [-0.06023481]
 [-0.15518121]]




+--------------------------+ 
| QR Factorization of A^tA | 
+--------------------------+ 

 The A^T A matrix: 
 [[37. 14.  8.]
 [14. 21. 14.]
 [ 8. 14. 33.]] 

 The QR factorization:

Q matrix:
[[-0.91672945  0.39755381  0.03947262]
 [-0.3468706  -0.74302954 -0.57235294]
 [-0.19821177 -0.53838468  0.81905679]]

R matrix:
[[-40.36087214 -22.8934597  -18.73101249]
 [  0.         -17.57525257

These factorizations are quite interesting. I really like the A matrix, and subsequent $Ax = b$, factorizations; unfortunately, the world doesn't usually work so cleanly.